Zend/zend_ast.c

```diff
@@ -449,14 +457,22 @@ ZEND_API int zend_ast_evaluate(zval *result, zend_ast *ast, zend_class_entry *sc
 			} else {
 				zval tmp;
 
-				zend_fetch_dimension_const(&tmp, &op1, &op2, (ast->attr == ZEND_DIM_IS) ? BP_VAR_IS : BP_VAR_R);
+				if (!Z_ISERROR(op1)) {
+					zend_fetch_dimension_const(&tmp, &op1, &op2, (ast->attr == ZEND_DIM_IS) ? BP_VAR_IS : BP_VAR_R);
 
-				if (UNEXPECTED(Z_ISREF(tmp))) {
-					ZVAL_DUP(result, Z_REFVAL(tmp));
+					if (Z_ISERROR(tmp) && (ast->attr != ZEND_DIM_PROPAGATE_ERROR)) {
+						ZVAL_NULL(&tmp);
+					}
+
+					if (UNEXPECTED(Z_ISREF(tmp))) {
+						ZVAL_DUP(result, Z_REFVAL(tmp));
+					} else {
+						ZVAL_DUP(result, &tmp);
+					}
+					zval_ptr_dtor(&tmp);
 				} else {
-					ZVAL_DUP(result, &tmp);
+					ZVAL_COPY(result, &op1);
 				}
```

Comment:
This code currently still propagates the error if it's an existing one, even if the propagate_error flag is not set. This is also why you're seeing the failure if you drop the ERROR check in update_constant. I'd suggest moving the check replacing ERROR with NULL further down:

```
				if (!Z_ISERROR(op1)) {
					zend_fetch_dimension_const(&tmp, &op1, &op2, (ast->attr == ZEND_DIM_IS) ? BP_VAR_IS : BP_VAR_R);
					if (UNEXPECTED(Z_ISREF(tmp))) {
						ZVAL_DUP(result, Z_REFVAL(tmp));
					} else {
						ZVAL_DUP(result, &tmp);
					}
					zval_ptr_dtor(&tmp);
				} else {
					ZVAL_ERROR(result);
				}

				if (Z_ISERROR(result && ast->attr != ZEND_DIM_PROPAGATE_ERROR) {
					ZVAL_NULL(result);
				}
```